### PR TITLE
dispatcher: allow PrivateUse1 backends to opt out of CompositeExplicitAutograd decomposition

### LINKED ISSUE #187576: Allow PrivateUse1 backends to opt out of CompositeExplicitAutograd decomposition
## Problem

The PyTorch dispatch table resolves `CompositeExplicitAutograd` (CEA) kernels **before** the backend fallback.  PrivateUse1 backends that register a single catch-all fallback kernel never see composite ops — they only receive the already-decomposed primitives (e.g. `softmax` → `_softmax`, `convolution` → `convolution_overrideable`).

The resolution order today (`OperatorEntry::computeDispatchTableEntryWithDebug`):

```
1. Direct PrivateUse1 kernel           ← wins
2a. CEANF alias kernel                 ← applied before fallback
2b. CEA alias kernel                   ← applied before fallback
2c. CIA / Autograd alias kernels
3. Backend fallback                    ← only reached if no alias kernel
```

For **proxy backends** like [SkyTorch](https://github.com/astefanutti/skytorch) that forward ATen ops verbatim to a remote GPU server, receiving decomposed ops breaks the RPC protocol.  The server receives `_softmax` instead of `softmax`, requiring it to handle every decomposed primitive separately.

The only workaround today is to register every CEA-decomposed op explicitly at the `PrivateUse1` dispatch key.

## Scale

- **538** ATen op overloads have `CompositeExplicitAutograd` registrations in `native_functions.yaml`.
- SkyTorch currently registers **280+** ops manually to work around this.
- Each new PyTorch release can silently introduce new CEA decompositions that break the backend with no warning.

## Ask

Add a backend-level mechanism for PrivateUse1 backends to opt out of CEA decomposition, so that ops with only a CEA kernel route to the registered backend fallback instead of being silently decomposed.

## Proposed API

```python
# Python
handle = torch.utils.skip_cea_decomposition_for_privateuse1()
# Active while handle is alive; releasing it restores previous behavior.
del handle
```

Internally backed by `c10::Dispatcher::setPrivateUse1SkipCEA()` which:
1. Sets an atomic flag on the dispatcher.
2. Refreshes all `PrivateUse1` dispatch table entries (force-recomputes using the new resolution order where CEA alias lookup is skipped for `PrivateUse1`).
3. Returns an RAII handle for clean teardown.

With this flag set, the resolution order becomes:

```
1. Direct PrivateUse1 kernel           ← wins (unchanged)
2a. CEANF alias                        ← skipped for PrivateUse1
2b. CEA alias                          ← skipped for PrivateUse1
3. Backend fallback                    ← now reached for CEA-only ops
```

## Affected components

- `aten/src/ATen/core/dispatch/Dispatcher.h` / `.cpp` — new `setPrivateUse1SkipCEA()` API
- `aten/src/ATen/core/dispatch/OperatorEntry.cpp` — skip CEA/CEANF alias lookup when flag is set
- `torch/csrc/Module.cpp` — Python binding `torch._C._dispatch_set_privateuse1_skip_cea`
- `torch/utils/backend_registration.py` — `torch.utils.skip_cea_decomposition_for_privateuse1()`

### PR DESCRIPTION
**Pull request by vishalgoyal316**
## Summary

Fixes #187576

The PyTorch dispatch table resolves `CompositeExplicitAutograd` (CEA) kernels **before** the backend fallback, so PrivateUse1 backends using a catch-all fallback never see composite ops — they only receive decomposed primitives (`softmax` → `_softmax`, `convolution` → `convolution_overrideable`, etc.).

For proxy backends like [SkyTorch](https://github.com/astefanutti/skytorch) (PyTorch with remote GPUs) that need to forward ATen ops verbatim to a remote server, this silent decomposition is the core problem: **538 ops** have CEA registrations and SkyTorch currently registers **280+** ops manually to work around it. Every PyTorch release can silently introduce new CEA decompositions that break the backend.

This PR adds a one-line opt-out:

```python
handle = torch.utils.skip_cea_decomposition_for_privateuse1()
# All CEA-only ops now route to the PrivateUse1 fallback verbatim.
# Release handle to restore previous behavior (RAII).
del handle
```

### How it works

**Before** (current resolution order for PrivateUse1):
```
1. Direct PrivateUse1 kernel     -- wins
2a. CEANF alias kernel           -- silently decomposes the op
2b. CEA alias kernel             -- silently decomposes the op
3. Backend fallback              -- never reached for CEA-covered ops
```

**After** (with skip enabled):
```
1. Direct PrivateUse1 kernel     -- wins (unchanged)
2a. CEANF alias                  -- SKIPPED for PrivateUse1
2b. CEA alias                    -- SKIPPED for PrivateUse1
3. Backend fallback              -- now reached for CEA-only ops
```

Direct `PrivateUse1` kernel registrations are completely unaffected.

### Implementation

| File | Change |
|---|---|
| `aten/src/ATen/core/dispatch/Dispatcher.h` | New `setPrivateUse1SkipCEA()` + `privateuse1SkipCEA()` + `std::atomic<bool>` field |
| `aten/src/ATen/core/dispatch/Dispatcher.cpp` | Implement `setPrivateUse1SkipCEA()`: set atomic, refresh all PrivateUse1 dispatch entries, return RAII handle |
| `aten/src/ATen/core/dispatch/OperatorEntry.cpp` | In `computeDispatchTableEntryWithDebug`: skip CEANF/CEA alias lookup when `dispatcher.privateuse1SkipCEA()` is set |
| `torch/csrc/Module.cpp` | `_dispatch_set_privateuse1_skip_cea` and `_dispatch_privateuse1_skip_cea_enabled` Python bindings |
| `torch/utils/backend_registration.py` | `skip_cea_decomposition_for_privateuse1()` public Python API |
| `torch/utils/__init__.py` | Re-export the new function |
| `test/test_privateuse1_skip_cea.py` | New test: RAII semantics, double-registration error, CEA op routing, direct-kernel unaffected |

## Test plan

```bash
python test/test_privateuse1_skip_cea.py -v
```

Tests cover:
- Flag is `False` initially
- RAII handle sets flag on enter and restores on `del`
- Double registration raises `RuntimeError("already registered")`
- A known CEA op (`aten::abs`) reaches the fallback (not CEA decomposition) when skip is enabled
- A direct `PrivateUse1` kernel registration is unaffected by the skip flag

Authored with AI assistant (Sonnet 4.6).

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by vishalgoyal316** [comment]
Closing - PR filed prematurely, not ready for review.
